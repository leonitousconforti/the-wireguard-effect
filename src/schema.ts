import * as Schema from "@effect/schema/Schema";
import * as Data from "effect/Data";

/** Generic wireguard interface config error. */
export class WireguardError extends Data.TaggedError("WireguardError")<{ message: string }> {}

/** @see https://github.com/WireGuard/wgctrl-go/blob/925a1e7659e675c94c1a659d39daa9141e450c7d/wgtypes/types.go#L236-L276 */
export class WireguardPeer extends Schema.Class<WireguardPeer>()({
    /**
     * The value for this is a decimal-string integer corresponding to the
     * persistent keepalive interval. The value 0 disables it.
     */
    PersistentKeepaliveInterval: Schema.number,

    /**
     * The value for this is IP/cidr, indicating a new added allowed IP entry
     * for the previously added peer entry. If an identical value already exists
     * as part of a prior peer, the allowed IP entry will be removed from that
     * peer and added to this peer.
     */
    AllowedIPs: Schema.array(Schema.string),

    /**
     * The value for this key is either IP:port for IPv4 or [IP]:port for IPv6,
     * indicating the endpoint of the previously added peer entry.
     */
    Endpoint: Schema.string.pipe(Schema.nonEmpty()),

    /**
     * The value for this key should be a lowercase hex-encoded public key of a
     * new peer entry.
     */
    PublicKey: Schema.string.pipe(Schema.nonEmpty()),
}) {}

/** @see https://github.com/WireGuard/wgctrl-go/blob/925a1e7659e675c94c1a659d39daa9141e450c7d/wgtypes/types.go#L207-L232 */
export class WireguardInterface extends Schema.Class<WireguardInterface>()({
    /**
     * The value for this is a decimal-string integer corresponding to the
     * listening port of the interface.
     */
    ListenPort: Schema.number,

    /**
     * The value for this is a decimal-string integer corresponding to the
     * fwmark of the interface. The value may 0 in the case of a set operation,
     * in which case it indicates that the fwmark should be removed.
     */
    FirewallMark: Schema.optional(Schema.number),

    /**
     * The value for this key should be a lowercase hex-encoded private key of
     * the interface. The value may be an all zero string in the case of a set
     * operation, in which case it indicates that the private key should be
     * removed.
     */
    PrivateKey: Schema.string,

    /**
     * Indicates that the subsequent peers (perhaps an empty list) should
     * replace any existing peers, rather than append to the existing peer
     * list.
     */
    ReplacePeers: Schema.optional(Schema.boolean, { default: () => true }),

    /** List of peers to add. */
    Peers: Schema.nonEmptyArray(WireguardPeer),
}) {}
