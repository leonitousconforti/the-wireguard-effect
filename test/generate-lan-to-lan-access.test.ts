import { describe, expect, it } from "@effect/vitest";

import * as Schema from "@effect/schema/Schema";
import * as Effect from "effect/Effect";
import * as WireguardConfig from "the-wireguard-effect/WireguardConfig";
import * as GenerateExample from "../examples/generate-lan-to-lan-access.js";

describe("wireguard generate lan to lan access", () => {
    it.effect("Should generate configs", () =>
        Effect.gen(function* () {
            const configs = yield* GenerateExample.program();
            expect(configs).toHaveLength(2);

            const config0 = yield* Schema.encode(WireguardConfig.WireguardConfig)(configs[0]);
            const config1 = yield* Schema.encode(WireguardConfig.WireguardConfig)(configs[1]);

            const keyMatcher = expect.stringMatching(/^[\d+/A-Za-z]{42}[048AEIMQUYcgkosw]=$/);
            const peerEntryMatcher = { PublicKey: keyMatcher, PresharedKey: keyMatcher };

            expect(config0).toMatchSnapshot({ PrivateKey: keyMatcher, Peers: [peerEntryMatcher] });
            expect(config1).toMatchSnapshot({ PrivateKey: keyMatcher, Peers: [peerEntryMatcher] });
        })
    );
});
