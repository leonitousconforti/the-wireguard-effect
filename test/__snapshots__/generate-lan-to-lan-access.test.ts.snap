// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wireguard generate lan to lan access > Should generate configs 1`] = `
{
  "Address": "10.0.0.1/24",
  "ListenPort": 51820,
  "Peers": [
    {
      "AllowedIPs": Set {
        "10.0.0.2/32",
        "192.168.2.1/24",
      },
      "Endpoint": "server2.wireguard.com:51821:51821",
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
  ],
  "PrivateKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
}
`;

exports[`wireguard generate lan to lan access > Should generate configs 2`] = `
{
  "Address": "10.0.0.2/24",
  "ListenPort": 51821,
  "Peers": [
    {
      "AllowedIPs": Set {
        "10.0.0.1/32",
        "192.168.1.1/24",
      },
      "Endpoint": "server1.wireguard.com:51820:51820",
      "PresharedKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
      "PublicKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
    },
  ],
  "PrivateKey": StringMatching /\\^\\[\\\\d\\+/A-Za-z\\]\\{42\\}\\[048AEIMQUYcgkosw\\]=\\$/,
}
`;
